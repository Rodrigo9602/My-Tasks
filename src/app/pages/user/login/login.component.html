<div class="content-wrapper">
    <div class="form-wrapper">
        <p class="header">Login</p>
        <form #customform="ngForm" class="custom-form" id="login-form" (ngSubmit)="onSubmit()">            
            <div class="inputs-container">
                <div class="input-wrapper">
                    <input type="email" class="input" placeholder="email" name="userEmail"
                        pattern="^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$" #userEmail="ngModel"
                        [style.border]="!userEmail.valid && userEmail.touched ? '1px solid red':'1px solid #fff' "
                        [(ngModel)]="email" required>
                    <div class="input-icon">
                        <fa-icon [icon]="emailIcon" />
                    </div>
                </div>
    
                <div class="input-wrapper">
                    <input type="{{ showPassword ? 'text' : 'password' }}" class="input" placeholder="password"
                        name="userPassword" pattern="^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)(?=.*[^\w\d\s:])([^\s]){8,}$"
                        [style.border]="!userPassword.valid && userPassword.touched  ? '1px solid red':'1px solid #fff' "
                        #userPassword="ngModel" [(ngModel)]="password" required>
                    <div class="input-icon">
                        <fa-icon [icon]="passIcon" />
                    </div>
                    <button type="button" class="btn-icon" (click)="onShowPassword()">
                        <fa-icon [icon]="showIcon" />
                    </button>
                   
                </div>
            </div>           

            <button type="submit" class="btn-custom" [disabled]="customform.invalid">Sign in</button>
        </form>

        <div class="redirection-wrapper">
            <p>You are not register yet? Join the family
                <a class="redirection-link" [routerLink]="['/register']">here</a>
            </p>
        </div>
    </div>
</div>