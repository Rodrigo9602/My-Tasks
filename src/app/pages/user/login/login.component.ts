import { Component, OnInit, OnDestroy } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SessionService } from '../../../services/session.service';
import { HttpClientModule } from '@angular/common/http';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { FormsModule } from '@angular/forms';
import { RouterModule, Router } from '@angular/router';

import { faLock, faEnvelope, faEye, faEyeSlash } from '@fortawesome/free-solid-svg-icons';
import { Toast } from '../../../global/toast.global';
import { Subscription } from 'rxjs';


@Component({
  selector: 'app-login',
  standalone: true,
  imports: [CommonModule, HttpClientModule, FontAwesomeModule, FormsModule, RouterModule],
  providers: [],
  templateUrl: './login.component.html',
  styleUrl: './login.component.scss'
})
export class LoginComponent implements OnInit, OnDestroy {

  public emailIcon = faEnvelope;
  public passIcon = faLock;
  public showIcon = faEye;

  public email: string = '';
  public password: string = '';

  public showPassword: boolean = false;
  private suscription: Subscription | undefined;


  constructor(private _sessionService: SessionService, private _router: Router) { }


  ngOnInit(): void {
    // clear localstorage
    localStorage.clear();
  }

  onSubmit() {
    // check if the input data is valid
    const emailRegex = new RegExp(/^([a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,})$/);
    const passwordRegex = new RegExp(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*]).{8,}$/);


    if (!emailRegex.test(this.email) && !passwordRegex.test(this.password)) {
      Toast.fire({
        icon: 'error',
        title: 'Email and password are invalid',
        text: 'Password must have at least: 8 characters, 1 uppercase letter, 1 lowercase letter, 1 number and 1 special character',
        timer: 3000,
      });
    } else if (!passwordRegex.test(this.password)) {
      Toast.fire({
        icon: 'error',
        title: 'Invalid password',
        text: 'Password must have at least: 8 characters, 1 uppercase letter, 1 lowercase letter, 1 number and 1 special character',
        timer: 3000,
      });
    } else if (!emailRegex.test(this.email)) {
      Toast.fire({
        icon: 'error',
        title: 'Invalid email',
      });
    } else {

      this.suscription = this._sessionService.login(this.email, this.password).subscribe({
        next: res => {
          if (res.status === 401) {
            Toast.fire({
              icon: 'error',
              title: res.options.message
            });
          } else {
            this._sessionService.setSession(res.id, res.access_token);
            // show welcome alert
            Toast.fire({
              icon: 'success',
              title: 'Welcome back!'
            });
            // redirect to home page
            this._router.navigate(['/home']);
          }
        },
      });
    }

  }

  onShowPassword() {
    this.showPassword = !this.showPassword;
    this.showPassword ? this.showIcon = faEyeSlash : this.showIcon = faEye;
  }

  ngOnDestroy(): void {
    this.suscription?.unsubscribe();
  }
}
