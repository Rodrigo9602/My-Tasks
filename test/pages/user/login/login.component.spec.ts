import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LoginComponent } from '../../../../src/app/pages/user/login/login.component';
import { ActivatedRoute } from '@angular/router';


describe('LoginComponent', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [LoginComponent],
      providers: [
        { provide: ActivatedRoute, useValue: {snapshot: {}}} 
      ]
    })
    .compileComponents();
    
    fixture = TestBed.createComponent(LoginComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match initial snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });
});
