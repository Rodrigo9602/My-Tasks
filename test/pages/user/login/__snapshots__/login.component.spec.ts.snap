// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should match initial snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="content-wrapper"
  >
    <div
      class="form-wrapper"
    >
      <p
        class="header"
      >
        Login
      </p>
      <form
        class="custom-form ng-untouched ng-pristine ng-valid"
        id="login-form"
        novalidate=""
      >
        <div
          class="inputs-container"
        >
          <div
            class="input-wrapper"
          >
            <input
              class="input ng-untouched ng-pristine ng-valid"
              name="userEmail"
              placeholder="email"
              required=""
              style="border: 1px solid #fff;"
              type="email"
            />
            <div
              class="input-icon"
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-envelope"
                  data-icon="envelope"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4L236.8 313.6c11.4 8.5 27 8.5 38.4 0L492.8 150.4c12.1-9.1 19.2-23.3 19.2-38.4c0-26.5-21.5-48-48-48H48zM0 176V384c0 35.3 28.7 64 64 64H448c35.3 0 64-28.7 64-64V176L294.4 339.2c-22.8 17.1-54 17.1-76.8 0L0 176z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </div>
          </div>
          <div
            class="input-wrapper"
          >
            <input
              class="input ng-untouched ng-pristine ng-valid"
              name="userPassword"
              placeholder="password"
              required=""
              style="border: 1px solid #fff;"
              type="password"
            />
            <div
              class="input-icon"
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-lock"
                  data-icon="lock"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M144 144v48H304V144c0-44.2-35.8-80-80-80s-80 35.8-80 80zM80 192V144C80 64.5 144.5 0 224 0s144 64.5 144 144v48h16c35.3 0 64 28.7 64 64V448c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V256c0-35.3 28.7-64 64-64H80z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </div>
            <button
              class="btn-icon"
              type="button"
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-eye"
                  data-icon="eye"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </button>
          </div>
        </div>
        <button
          class="btn-custom"
          type="submit"
        >
          Sign in
        </button>
      </form>
      <div
        class="redirection-wrapper"
      >
        <p>
          You are not register yet? Join the family 
          <a
            class="redirection-link"
            href="/register"
          >
            here
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
